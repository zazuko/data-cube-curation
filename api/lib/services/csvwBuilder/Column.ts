import { csvw } from '../../namespaces'
import * as Csvw from './index'
import TypedClownfaceEntity from '../../read-model/TypedClownfaceEntity'
import { namespace, literal, property } from '../../read-model/decorators'
import { Constructor } from '../../read-model/TypedEntityFactory'

export function ColumnMixin<TBase extends Constructor> (Base: TBase) {
  @namespace(csvw)
  abstract class Column extends Base implements Csvw.Column {
    @literal()
    public propertyUrl: string;

    @literal({ path: csvw.lang })
    public language: string;

    @property()
    public datatype: TypedClownfaceEntity;

    @property()
    public valueUrl: string;

    @literal()
    public title: string;

    @literal({ path: csvw.suppressOutput, type: Boolean })
    public suppressed: boolean;
  }

  return Column
}
