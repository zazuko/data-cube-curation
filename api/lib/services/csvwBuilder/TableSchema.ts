import { csvw } from '../../namespaces'
import TypedClownfaceEntity from '../../read-model/TypedClownfaceEntity'
import * as Csvw from './index'
import { literal, namespace } from '../../read-model/decorators'

@namespace(csvw)
export class TableSchema extends TypedClownfaceEntity implements Csvw.TableSchema {
  public set columns (columns: Csvw.Column[]) {
    this.addList(csvw.column, columns)
  }

  @literal()
  public aboutUrl: string;
}
