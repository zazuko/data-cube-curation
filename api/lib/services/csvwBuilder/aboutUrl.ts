import { Project } from '../../read-model'
import isUri = require('is-uri')

export function getAboutUrl (project: Project, uriTemplate: any) {
  uriTemplate.expressions.forEach(expression => {
    expression.params.forEach(p => {
      // TODO: required until grncdr/uri-template#19 is fixed
      p.explode = ''
    })
  })

  if (isUri(uriTemplate.prefix)) {
    return uriTemplate.toString()
  }

  return project.baseUri + uriTemplate.toString()
}
