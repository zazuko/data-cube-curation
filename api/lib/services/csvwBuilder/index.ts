import { TypedClownfaceEntity } from '../../read-model/TypedClownfaceEntity'

export interface Column {
  suppressed: boolean;
  propertyUrl: string;
  language: string;
  datatype: TypedClownfaceEntity;
  valueUrl: string;
}

export interface TableSchema {
  columns: Column[];
  aboutUrl: string;
}

export interface Mapping {
  newColumn(col: { name: string }): Column;
  addDialect(): void;
  readonly tableSchema: TableSchema;
}
