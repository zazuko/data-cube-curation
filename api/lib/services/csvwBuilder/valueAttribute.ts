import { RdfResource } from '@tpluscode/rdfine'
import { literal } from '@rdfjs/data-model'
import { shrink } from '@zazuko/rdf-vocabularies'
import * as Table from '../../read-model/Table/index'
import * as Csvw from '@rdfine/csvw'
import { csvw, rdf } from '../../namespaces'

const xsdPrefix = /^xsd:(.+)/

function getBuiltInDatatypeName (datatype: RdfResource): string | null {
  switch (datatype.id.value) {
    case (csvw.JSON.value):
      return 'json'
    case (rdf.HTML.value):
      return 'html'
    case (rdf.XMLLiteral.value):
      return 'xml'
  }

  const shrunk = shrink(datatype.id.value)
  const matches = shrunk.match(xsdPrefix)

  if (!matches) {
    return null
  }

  switch (matches[1]) {
    case 'double':
      return 'number'
    case 'base64Binary':
      return 'binary'
    case 'dateTime':
      return 'datetime'
    case 'anyAtomicType':
      return 'any'
  }

  return matches[1]
}

export function valueAttributeToCsvwColumn (attribute: Table.ValueAttribute, csvwColumn: Csvw.Column) {
  csvwColumn.default = attribute.default

  if (attribute.language) {
    csvwColumn.language = attribute.language
  } else if (attribute.datatype) {
    const builtInType = getBuiltInDatatypeName(attribute.datatype)
    const hasParameters = attribute.parameters && Object.entries(attribute.parameters).length > 0

    if (builtInType && hasParameters) {
      const derivedDatatype = csvwColumn._create(csvwColumn._node.blankNode())
      derivedDatatype[csvw.base.value] = literal(builtInType)

      Object.entries(attribute.parameters)
        .forEach(([param, value]) => {
          derivedDatatype[csvw(param).value] = literal(value)
        })

      csvwColumn.datatype = derivedDatatype
    } else {
      csvwColumn.datatype = attribute.datatype
    }
  }

  return csvwColumn
}
