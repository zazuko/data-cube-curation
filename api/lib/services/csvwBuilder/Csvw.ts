import { csvw, rdf } from '../../namespaces'
import * as Csvw from './index'
import TypedClownfaceEntity from '../../read-model/TypedClownfaceEntity'
import { TableSchema } from './TableSchema'
import { Term } from 'rdf-js'
import { factory } from '../../read-model/TypedEntityFactory'
import { ColumnMixin } from './Column'

export default class extends TypedClownfaceEntity implements Csvw.Mapping {
  public constructor (o: { dataset: any; term: Term }) {
    super(o)

    this.addOut(rdf.type, csvw.CsvwMapping)
  }

  public get tableSchema () {
    let node = this.out(csvw.tableSchema)
    if (!node.term) {
      this.addOut(csvw.tableSchema, newNode => {
        node = newNode
      })
    }

    return new TableSchema(node)
  }

  public addDialect () {
    this.addOut(csvw.dialect, dialect => {
      dialect.addOut(csvw.header, true)
      dialect.addOut(csvw.delimiter, ';')
      dialect.addOut(csvw.quoteChar, '"')
    })
  }

  public newColumn (col: { name: string }) {
    const node = this.blankNode().addOut(csvw.title, col.name)

    return factory.createEntity<Csvw.Column>(node, [ ColumnMixin ])
  }
}
