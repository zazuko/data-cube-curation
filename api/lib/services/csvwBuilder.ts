import $rdf from 'rdf-ext'
import { Dataset } from 'rdf-js'
import cf from 'clownface'
import { valueAttributeToCsvwColumn } from './csvwBuilder/valueAttribute'
import { referenceAttributeToCsvwColumn } from './csvwBuilder/referenceAttribute'
import { error } from '../log'
import CsvwGraph from './csvwBuilder/Csvw'
import * as Table from '../read-model/Table'
import { BaseTable } from '../read-model/Table/Table'
import * as Csvw from './csvwBuilder/index'
import { getAbsoluteUrl } from './csvwBuilder/aboutUrl'
import parser = require('uri-template')

type Attribute = Table.ReferenceAttribute | Table.ValueAttribute | Table.Attribute

function createCsvwColumn (csvwGraph: Csvw.Mapping, table: Table.Table, attribute: Attribute): Csvw.Column | null {
  let csvwColumn: Csvw.Column | null = null

  if ('column' in attribute) {
    csvwColumn = csvwGraph.newColumn({
      name: attribute.column.name,
    })

    csvwColumn = valueAttributeToCsvwColumn(attribute, csvwColumn)
  }

  if ('columnMappings' in attribute) {
    const column = attribute.columnMappings[0].sourceColumn
    csvwColumn = csvwGraph.newColumn({
      name: column.name,
    })

    csvwColumn = referenceAttributeToCsvwColumn(attribute, csvwColumn)
  }

  if (csvwColumn) {
    const propertyTemplate = attribute.propertyTemplate
    const parsed = parser.parse(propertyTemplate)
    csvwColumn.propertyUrl = getAbsoluteUrl(table.project, parsed)
    return csvwColumn
  }

  error(`The types of <%s> did not contain one of the expected datacube:Attribute types`, attribute.id)
  return null
}

export function buildCsvw (tableOrDataset: Table.Table | Table.DimensionTable | { dataset: Dataset; tableId: string }) {
  let table: Table.Table | Table.DimensionTable
  if ('dataset' in tableOrDataset) {
    table = BaseTable.factory.createEntity(cf({
      dataset: tableOrDataset.dataset,
      term: $rdf.namedNode(tableOrDataset.tableId),
    }), [BaseTable])
  } else {
    table = tableOrDataset
  }

  const csvwGraph = new CsvwGraph({ dataset: $rdf.dataset(), term: $rdf.namedNode(`${table.id.value}/csvw`) })

  csvwGraph.addDialect()

  if ('identifierTemplate' in table) {
    const parsed = parser.parse(table.identifierTemplate)
    csvwGraph.tableSchema.aboutUrl = getAbsoluteUrl(table.project, parsed)
  }

  const attributes: Attribute[] = table.attributes
  const mappedColumns = attributes.reduce((columns, attr) => {
    const column = createCsvwColumn(csvwGraph, table, attr)

    if (column) {
      columns.push(column)
    }

    return columns
  }, [] as Csvw.Column[])

  const suppressedColumns = table.columns.reduce((mapped, column: Table.Column) => {
    if (!mappedColumns.find(c => c.title === column.name)) {
      const csvwColumn = csvwGraph.newColumn({
        name: column.name,
      })
      csvwColumn.suppressed = true

      mapped.push(csvwColumn)
    }

    return mapped
  }, [] as Csvw.Column[])

  csvwGraph.tableSchema.columns = [...mappedColumns, ...suppressedColumns]

  return csvwGraph
}
