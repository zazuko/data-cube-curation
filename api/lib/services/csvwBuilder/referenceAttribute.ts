import parser from 'uri-template'
import { error, warning } from '../../log'
import * as Table from '../../read-model/Table'
import * as Csvw from '../csvwBuilder/index'
import { getAbsoluteUrl } from './aboutUrl'

export function referenceAttributeToCsvwColumn (attribute: Table.ReferenceAttribute, csvwColumn: Csvw.Column) {
  const referencedTable = attribute.referencedTable

  const columnNameMap = attribute.columnMappings
    .reduce((map, mapping) => {
      const from = mapping.sourceColumn.name
      const to = mapping.referencedColumn.name

      map.set(to, from)

      return map
    }, new Map<string, string>())

  if (typeof referencedTable.identifierTemplate === 'string') {
    const uriTemplate = parser.parse(referencedTable.identifierTemplate)
    uriTemplate.expressions.forEach(expression => {
      expression.params.forEach(p => {
        if (columnNameMap.has(p.name)) {
          p.name = columnNameMap.get(p.name)
        } else {
          warning('Column name %s was not found in template for table <%s>', p.name, referencedTable.id)
        }
      })
    })

    csvwColumn.valueUrl = getAbsoluteUrl(referencedTable.project, uriTemplate)
  }

  error(`Failed to create column for reference attribute <%s>`, attribute.id)
  return csvwColumn
}
