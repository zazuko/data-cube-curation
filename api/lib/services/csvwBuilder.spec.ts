import { buildCsvw } from './csvwBuilder'
import { Column, DimensionTable, ValueAttribute } from '../read-model/Table'
import { namedNode } from '@rdfjs/data-model'
import { dataCube, xsd } from '../namespaces'
import * as specGraphs from './csvwBuilder.spec-graphs'

type RecursivePartial<T> = {
  [P in keyof T]?: RecursivePartial<T[P]>;
};

const { ids } = specGraphs

describe('csvwBuilder', () => {
  describe('mapping for FactTable', () => {
    it('includes unmapped source columns as suppressed', async () => {
      // given
      const dataset = await specGraphs.unmappedColumnGraph()

      // when
      const csvwDataset = buildCsvw({ dataset, tableId: ids.tableId })

      // then
      expect(csvwDataset._node.dataset.toCanonical()).toMatchSnapshot()
    })

    it('maps attribute', async () => {
      // given
      const dataset = await specGraphs.mappedColumnGraph()

      // when
      const csvwDataset = buildCsvw({ dataset, tableId: ids.tableId })

      // then
      expect(csvwDataset._node.dataset.toCanonical()).toMatchSnapshot()
    })

    it('maps multiple attributes mapping same column', async () => {
      // given
      const dataset = await specGraphs.multipleMappedColumnsGraph()

      // when
      const csvwDataset = buildCsvw({ dataset, tableId: ids.tableId })

      // then
      expect(csvwDataset._node.dataset.toCanonical()).toMatchSnapshot()
    })

    it('maps attribute with datatype', async () => {
      // given
      const dataset = await specGraphs.columnMappedWithDatatypeGraph()

      // when
      const csvwDataset = buildCsvw({ dataset, tableId: ids.tableId })

      // then
      expect(csvwDataset._node.dataset.toCanonical()).toMatchSnapshot()
    })

    it('maps attribute with datatype parameters', async () => {
      // given
      const dataset = await specGraphs.columnMappedWithDatatypeAndParamsGraph()

      // when
      const csvwDataset = buildCsvw({ dataset, tableId: ids.tableId })

      // then
      expect(csvwDataset._node.dataset.toCanonical()).toMatchSnapshot()
    })

    it('maps attribute with language tag', async () => {
      // given
      const dataset = await specGraphs.columnMappedWithLanguageGraph()

      // when
      const csvwDataset = buildCsvw({ dataset, tableId: ids.tableId })

      // then
      expect(csvwDataset._node.dataset.toCanonical()).toMatchSnapshot()
    })

    it('maps reference attribute', async () => {
      // given
      const dataset = await specGraphs.referenceAttributeGraph()

      // when
      const csvwDataset = buildCsvw({ dataset, tableId: `http://reference-attribute.test/fact-table` })

      // then
      expect(csvwDataset._node.dataset.toCanonical()).toMatchSnapshot()
    })
  })

  describe('mapping for DimensionTable', () => {
    it('uses absolute identifierTemplate for aboutUrl', async () => {
      // given
      const table: RecursivePartial<DimensionTable> = {
        id: namedNode('http://reference-attribute.test/fact-table'),
        identifierTemplate: 'http://example.com/{foo}/{bar}',
        columns: [],
        attributes: [],
        types: [
          dataCube.DimensionTable,
        ],
        project: {
          baseUri: 'http://example.com/tst-project',
        },
      }

      // when
      const csvwDataset = buildCsvw(table as any)

      // then
      expect(csvwDataset.tableSchema.aboutUrl).toEqual('http://example.com/{foo}/{bar}')
    })

    it('concatenates project base when identifier template is note absolute', async () => {
      // given
      const table: RecursivePartial<DimensionTable> = {
        id: namedNode('http://reference-attribute.test/fact-table'),
        identifierTemplate: 'table/{foo}/{bar}',
        columns: [],
        attributes: [],
        types: [
          dataCube.DimensionTable,
        ],
        project: {
          baseUri: 'http://example.com/tst-project/',
        },
      }

      // when
      const csvwDataset = buildCsvw(table as any)

      // then
      expect(csvwDataset.tableSchema.aboutUrl).toEqual('http://example.com/tst-project/table/{foo}/{bar}')
    })

    it('uses base URI for relative properties', async () => {
      // given
      const column: Partial<Column> = {
        id: namedNode('http://reference-attribute.test/column'),
        name: 'column',
      }
      const attribute: RecursivePartial<ValueAttribute> = {
        id: namedNode('http://reference-attribute.test/attribute'),
        column,
        datatype: { id: xsd.string },
        propertyTemplate: 'property/{foo}/{bar}',
      }
      const table: RecursivePartial<DimensionTable> = {
        id: namedNode('http://reference-attribute.test/fact-table'),
        identifierTemplate: 'table/{foo}/{bar}',
        columns: [column],
        attributes: [attribute],
        types: [
          dataCube.DimensionTable,
        ],
        project: {
          baseUri: 'http://example.com/tst-project/',
        },
      }

      // when
      const csvwDataset = buildCsvw(table as any)

      // then
      expect(csvwDataset.tableSchema.columns[0].propertyUrl).toEqual('http://example.com/tst-project/property/{foo}/{bar}')
    })

    it('uses concrete rdf:property if template is missing', async () => {
      // given
      const column: Partial<Column> = {
        id: namedNode('http://reference-attribute.test/column'),
        name: 'column',
      }
      const attribute: RecursivePartial<ValueAttribute> = {
        id: namedNode('http://reference-attribute.test/attribute'),
        column,
        datatype: { id: xsd.string },
        propertyTemplate: '',
        predicate: 'http://example.com/prop',
      }
      const table: RecursivePartial<DimensionTable> = {
        id: namedNode('http://reference-attribute.test/fact-table'),
        identifierTemplate: 'table/{foo}/{bar}',
        columns: [column],
        attributes: [attribute],
        types: [
          dataCube.DimensionTable,
        ],
        project: {
          baseUri: 'http://example.com/tst-project/',
        },
      }

      // when
      const csvwDataset = buildCsvw(table as any)

      // then
      expect(csvwDataset.tableSchema.columns[0].propertyUrl).toEqual('http://example.com/prop')
    })
  })
})
