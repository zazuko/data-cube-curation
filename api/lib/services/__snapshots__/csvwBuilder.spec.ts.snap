// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`csvwBuilder includes unmapped source columns as suppressed 1`] = `
"<http://example.com/table/Observation/csvw> <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> <http://www.w3.org/ns/csvw#CsvwMapping> .
<http://example.com/table/Observation/csvw> <http://www.w3.org/ns/csvw#dialect> _:c14n3 .
<http://example.com/table/Observation/csvw> <http://www.w3.org/ns/csvw#tableSchema> _:c14n0 .
_:c14n0 <http://www.w3.org/ns/csvw#column> _:c14n2 .
_:c14n1 <http://www.w3.org/ns/csvw#suppressOutput> \\"true\\"^^<http://www.w3.org/2001/XMLSchema#boolean> .
_:c14n1 <http://www.w3.org/ns/csvw#title> \\"station_name\\" .
_:c14n2 <http://www.w3.org/1999/02/22-rdf-syntax-ns#first> _:c14n1 .
_:c14n2 <http://www.w3.org/1999/02/22-rdf-syntax-ns#rest> <http://www.w3.org/1999/02/22-rdf-syntax-ns#nil> .
_:c14n3 <http://www.w3.org/ns/csvw#delimiter> \\";\\" .
_:c14n3 <http://www.w3.org/ns/csvw#header> \\"true\\"^^<http://www.w3.org/2001/XMLSchema#boolean> .
_:c14n3 <http://www.w3.org/ns/csvw#quoteChar> \\"\\\\\\"\\" .
"
`;

exports[`csvwBuilder maps attribute 1`] = `
"<http://example.com/table/Observation/csvw> <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> <http://www.w3.org/ns/csvw#CsvwMapping> .
<http://example.com/table/Observation/csvw> <http://www.w3.org/ns/csvw#dialect> _:c14n3 .
<http://example.com/table/Observation/csvw> <http://www.w3.org/ns/csvw#tableSchema> _:c14n0 .
_:c14n0 <http://www.w3.org/ns/csvw#column> _:c14n2 .
_:c14n1 <http://www.w3.org/ns/csvw#propertyUrl> \\"http://schema.org/name\\" .
_:c14n1 <http://www.w3.org/ns/csvw#title> \\"station_name\\" .
_:c14n2 <http://www.w3.org/1999/02/22-rdf-syntax-ns#first> _:c14n1 .
_:c14n2 <http://www.w3.org/1999/02/22-rdf-syntax-ns#rest> <http://www.w3.org/1999/02/22-rdf-syntax-ns#nil> .
_:c14n3 <http://www.w3.org/ns/csvw#delimiter> \\";\\" .
_:c14n3 <http://www.w3.org/ns/csvw#header> \\"true\\"^^<http://www.w3.org/2001/XMLSchema#boolean> .
_:c14n3 <http://www.w3.org/ns/csvw#quoteChar> \\"\\\\\\"\\" .
"
`;

exports[`csvwBuilder maps attribute with datatype 1`] = `
"<http://example.com/table/Observation/csvw> <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> <http://www.w3.org/ns/csvw#CsvwMapping> .
<http://example.com/table/Observation/csvw> <http://www.w3.org/ns/csvw#dialect> _:c14n3 .
<http://example.com/table/Observation/csvw> <http://www.w3.org/ns/csvw#tableSchema> _:c14n0 .
_:c14n0 <http://www.w3.org/ns/csvw#column> _:c14n2 .
_:c14n1 <http://www.w3.org/ns/csvw#datatype> <http://www.w3.org/2001/XMLSchema#TOKEN> .
_:c14n1 <http://www.w3.org/ns/csvw#propertyUrl> \\"http://schema.org/name\\" .
_:c14n1 <http://www.w3.org/ns/csvw#title> \\"station_name\\" .
_:c14n2 <http://www.w3.org/1999/02/22-rdf-syntax-ns#first> _:c14n1 .
_:c14n2 <http://www.w3.org/1999/02/22-rdf-syntax-ns#rest> <http://www.w3.org/1999/02/22-rdf-syntax-ns#nil> .
_:c14n3 <http://www.w3.org/ns/csvw#delimiter> \\";\\" .
_:c14n3 <http://www.w3.org/ns/csvw#header> \\"true\\"^^<http://www.w3.org/2001/XMLSchema#boolean> .
_:c14n3 <http://www.w3.org/ns/csvw#quoteChar> \\"\\\\\\"\\" .
"
`;

exports[`csvwBuilder maps attribute with language tag 1`] = `
"<http://example.com/table/Observation/csvw> <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> <http://www.w3.org/ns/csvw#CsvwMapping> .
<http://example.com/table/Observation/csvw> <http://www.w3.org/ns/csvw#dialect> _:c14n3 .
<http://example.com/table/Observation/csvw> <http://www.w3.org/ns/csvw#tableSchema> _:c14n0 .
_:c14n0 <http://www.w3.org/ns/csvw#column> _:c14n2 .
_:c14n1 <http://www.w3.org/ns/csvw#lang> \\"en\\" .
_:c14n1 <http://www.w3.org/ns/csvw#propertyUrl> \\"http://schema.org/name\\" .
_:c14n1 <http://www.w3.org/ns/csvw#title> \\"station_name\\" .
_:c14n2 <http://www.w3.org/1999/02/22-rdf-syntax-ns#first> _:c14n1 .
_:c14n2 <http://www.w3.org/1999/02/22-rdf-syntax-ns#rest> <http://www.w3.org/1999/02/22-rdf-syntax-ns#nil> .
_:c14n3 <http://www.w3.org/ns/csvw#delimiter> \\";\\" .
_:c14n3 <http://www.w3.org/ns/csvw#header> \\"true\\"^^<http://www.w3.org/2001/XMLSchema#boolean> .
_:c14n3 <http://www.w3.org/ns/csvw#quoteChar> \\"\\\\\\"\\" .
"
`;

exports[`csvwBuilder maps multiple attributes mapping same column 1`] = `
"<http://example.com/table/Observation/csvw> <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> <http://www.w3.org/ns/csvw#CsvwMapping> .
<http://example.com/table/Observation/csvw> <http://www.w3.org/ns/csvw#dialect> _:c14n5 .
<http://example.com/table/Observation/csvw> <http://www.w3.org/ns/csvw#tableSchema> _:c14n0 .
_:c14n0 <http://www.w3.org/ns/csvw#column> _:c14n1 .
_:c14n1 <http://www.w3.org/1999/02/22-rdf-syntax-ns#first> _:c14n2 .
_:c14n1 <http://www.w3.org/1999/02/22-rdf-syntax-ns#rest> _:c14n3 .
_:c14n2 <http://www.w3.org/ns/csvw#propertyUrl> \\"http://schema.org/name\\" .
_:c14n2 <http://www.w3.org/ns/csvw#title> \\"station_name\\" .
_:c14n3 <http://www.w3.org/1999/02/22-rdf-syntax-ns#first> _:c14n4 .
_:c14n3 <http://www.w3.org/1999/02/22-rdf-syntax-ns#rest> <http://www.w3.org/1999/02/22-rdf-syntax-ns#nil> .
_:c14n4 <http://www.w3.org/ns/csvw#lang> \\"fr\\" .
_:c14n4 <http://www.w3.org/ns/csvw#propertyUrl> \\"http://schema.org/name\\" .
_:c14n4 <http://www.w3.org/ns/csvw#title> \\"station_name\\" .
_:c14n5 <http://www.w3.org/ns/csvw#delimiter> \\";\\" .
_:c14n5 <http://www.w3.org/ns/csvw#header> \\"true\\"^^<http://www.w3.org/2001/XMLSchema#boolean> .
_:c14n5 <http://www.w3.org/ns/csvw#quoteChar> \\"\\\\\\"\\" .
"
`;
